### DISCUSSION TITLE
NextJS Dynamic Routing with Firebase

### DISCUSSION THREAD

**Discussion by dennemark** [firebase/firebase-js-sdk#4980] [General]
Edit: Consider reading this comment first: https://github.com/firebase/firebase-js-sdk/discussions/4980#discussioncomment-3703537

---
Hi,

I just have to share another solution for using NextJS with firebase, that I haven't found on the web yet.

I have been using `next export` to create a static site and host it. However, this approach is missing dynamic routing, leading to 404 pages, when urls are shared.
An alternative I have seen, is using a cloud function with custom nextjs server and routing, however it breaks my workflow with cloud functions and is therefore, not preferred.

Dynamic routing can be made possible easily however, by using firebase hosting rewrites. And instead of writing them on your own, you can easily copy pase them from your build nextjs page. Just go to `.next/routes-manifest.json` and copy the dynamic routes section. I.e.:
```
"dynamicRoutes": [  
        {
            "page": "/user/[id]",
            "regex": "^/user/([^/]+?)(?:/)?$",
            "routeKeys": {
                "id": "id"
            },
            "namedRegex": "^/user/(?<id>[^/]+?)(?:/)?$"
        }]
```
and paste it to `firebase.json` rewrites and edit as follows:

```
"rewrites": [
        {
            "destination": "/user/[id].html",  // ADD .html !  and rename page to destination
            "regex": "^/user/([^/]+?)(?:/)?$",
        }]
```

done!

Hope it will help some people.

Edit 03.05.2022:
If above rewrite does not work, maybe it works to add `/index` to the destination. Thanks to aki matsumoto for the hint!
```
"rewrites": [
        {
            "destination": "/user/[id]/index.html",  // ADD .html !  and rename page to destination
            "regex": "^/user/([^/]+?)(?:/)?$",
        }]
```

**Comment by stuartrobinson3007** [comment]
Thank you! <3

**Comment by samuelrondot** [comment]
Thank you ! I was searching for a solution since hours 🙏

**Reply by silverit** [reply]
Can you show me exactly how you do it? I tried changing the file firebase.json, but it's not working!!!
```
"rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      },
      {
        "destination": "/post/[slug].html",
        "regex": "^/post/([^/]+?)(?:/)?$"
      },
      {
        "destination": "/tag/[slug].html",
        "regex": "^/tag/([^/]+?)(?:/)?$"
      }
    ]
```

**Reply by JoppeMeijers** [reply]
@silverit  did you find any solution? Because i have the same issue as you with exact the same firebase.json

**Reply by silverit** [reply]
> @silverit did you find any solution? Because i have the same issue as you with exact the same firebase.json

In my case, I added this line `trailingSlash: true,` to file `next.config.js`. Let NextJS generate the `index.html` file for each slug. So I didn't need to edit the file `firebase.json`, I keep the file `firebase.json` as the code below:
```
{
  "hosting": {
    "public": "out",
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ]
  }
}

```

**Reply by samuelrondot** [reply]
> @silverit did you find any solution? Because i have the same issue as you with exact the same firebase.json

Do you use SSR ? If you do , then you should use Vercel instead of Firebase . SSR doesn’t work on Firebase

**Reply by hkropp** [reply]
@samuelrondot for SSR to work with firebase you also need to add a rewrite rule such as:

```
{
    "source": "/nextjsFunc",
    "function": "ssrPROJECTNAME"
}
```

Replace PROJECTNAME with your firebase project name.

**Comment by silverit** [comment]
Thanks for your help. I found a way to fix my issue. I just add this
line `trailingSlash: true` to file `next.config.js`. Let NextJS generate the index.html file for
each slug.

On Tue, May 3, 2022 at 8:42 PM aki matsumoto ***@***.***>
wrote:

> I made the following changes and it works.
>
> {-  "destination": "/post/[slug].html",+  "destination": "/post/[slug]/index.html",
>   "regex": "^/post/([^/]+?)(?:/)?$"
> }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/firebase/firebase-js-sdk/discussions/4980#discussioncomment-2680403>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARY36G6ENVJQS2OMCHVI6TVIEULVANCNFSM46CP4O7Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Thanks & B. Regards,

Phuong Vo

Mobile App Developer

m:

0976071270

e:

***@***.***

**Comment by prs99** [comment]
I have the same issue but this solution didn't work for me

**Reply by jeff-r-skillrev** [reply]
I am using Next.js to generate a static site. It's only partially working for me. Changing `trailingSlash: true,` definitely got me one step closer. Any routes that do have a dynamic `/users/[slug]` do not work, but routes that are static work now, such as `/users`.

I'll report back if I get a more comprehensive answer on this.

**Reply by jeff-r-skillrev** [reply]
After carefully re-reading comments here, it seems like the solution was already stated. Here is a summary of changes required for me:

- Add `trailingSlash: true,` to next.js.config
  - This lets all static routing work
- Add dynamic / "slug" routes to the rewrites section of firebase.json

Here is a gist of a utility script that can grab all of your dynamic slug routes from the .next folder, and place them in firebase.json automatically (mostly for CI related use cases).

https://gist.github.com/jeff-r-koyaltech/6195f546764443d0cc7c5efe48fbe896

**Reply by ibussieres** [reply]
@jeff-r-koyaltech I'm stuck at the same step you were initially stuck on. My static routes are now working thanks to the trailingSlash parameter.

But my dynamic routes still all rewrite my app root.
I've tried with `/resume/[resumeId]/index.html` and `/resume/[resumeId].html`. Any chance you know what I'm missing here ?

```
  "hosting": {
    "public": "out",
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      },
      {
        "destination": "/resume/[resumeId]/index.html",
        "regex": "^/resume/([^/]+?)(?:/)?$"
      }
    ]
  },
```

```
  "dynamicRoutes": [
    {
      "page": "/resume/[resumeId]",
      "regex": "^/resume/([^/]+?)(?:/)?$",
      "routeKeys": { "resumeId": "resumeId" },
      "namedRegex": "^/resume/(?<resumeId>[^/]+?)(?:/)?$"
    }
  ],
```

**Reply by jeff-r-skillrev** [reply]
hey @ibussieres , in firebase.json, I don't have an entry for:

```
     {
        "source": "**",
        "destination": "/index.html"
      },
```

I think it might be worth trying without that in case it's matching on things that it shouldn't. My gist file (above) doesn't produce that kind of entry.

**Reply by ibussieres** [reply]
OMG. That was it.
Thanks for the (really fast) reply ! I should've tried that, but I saw people leaving it there in the previous responses. Turns out the correct way is without the catch all.

**Reply by dominicdev** [reply]
@ibussieres  can you show me how you did it? im having the same issue in next 13

**Reply by jeff-r-skillrev** [reply]
@dominicdev try the gist / script above. If you let it fully generate your file, then it should work. That's what @ibussieres did.

**Reply by jeff-r-skillrev** [reply]
"I wrote a script [...]"

@haris-aqeel  I am confused. Your gist is identical to mine:
https://gist.github.com/jeff-r-koyaltech/6195f546764443d0cc7c5efe48fbe896

I am glad my script was helpful to you, but please don't take credit for someone else's work. Did you mean to improve it in some way? If so, great. Please share that.

**Reply by haris-aqeel** [reply]
@jeff-r-koyaltech I actually improved it to pass the path dynamically through cmd as args instead of writing it in the file, but I realized it was not a significant contribution, so removed it. No issues.

**Comment by benjaminbialy** [comment]
You're a legend!

**Comment by haris-aqeel** [comment]


**Reply by jeff-r-skillrev** [reply]
"I wrote a script [...]"

I am confused. Your gist is identical to mine:
https://gist.github.com/jeff-r-koyaltech/6195f546764443d0cc7c5efe48fbe896

I am glad my script was helpful to you, but please don't take credit for someone else's work.

**Comment by Manoj-nathwani** [comment]
I managed to solve this by doing the following:
- Loop through all html files in the `/out` dir and update the `rewrites` section in the `firebase.json` file
- Make sure `/` (the root dir) is at the very end of the file

Here's the code.... all made by chat-gpt of course 😅

## package.json
```
"scripts": {
  "dev": "next dev",
  "build": "next build && node updateFirebaseRewrites.js",
  "deploy": "npm run build && firebase deploy",
  "start": "next start",
  "lint": "next lint"
},
```

## updateFirebaseRewrites.js
```
const fs = require('fs');
const path = require('path');

const outDirectory = path.join(__dirname, 'out'); // Adjust if your directory structure is different
const firebaseConfigPath = path.join(__dirname, 'firebase.json');

// Function to read all .html files from a directory
function readHtmlFiles(dir) {
    const files = fs.readdirSync(dir);
    let htmlFiles = [];

    files.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);

        if (stat.isDirectory()) {
            // Recursively read nested directories
            htmlFiles = [...htmlFiles, ...readHtmlFiles(filePath)];
        } else if (file.endsWith('.html') && file !== 'index.html') {
            // Remove the 'out' directory and '.html' extension from path
            const relativePath = path.relative(outDirectory, filePath).replace(/\.html$/, '');
            htmlFiles.push(`/${relativePath}`);
        }
    });

    return htmlFiles;
}

const htmlFiles = readHtmlFiles(outDirectory);

// Generate rewrite rules for each HTML file
let rewrites = htmlFiles.map(file => ({
    source: `${file}{,/**}`,
    destination: `${file}.html`
}));

// Add the root path rewrite at the end
rewrites.push({
    source: "/{,/**}",
    destination: "/index.html"
});

// Read the existing firebase.json configuration
const firebaseConfig = JSON.parse(fs.readFileSync(firebaseConfigPath, 'utf8'));

// Update the rewrites in firebase.json
firebaseConfig.hosting = firebaseConfig.hosting || {};
firebaseConfig.hosting.rewrites = rewrites;

// Write the updated configuration back to firebase.json
fs.writeFileSync(firebaseConfigPath, JSON.stringify(firebaseConfig, null, 2));

console.log('firebase.json has been updated with new rewrites.');

```

**Comment by onurays** [comment]
Perfect, thank you 🙌

**Comment by ssh-randy** [comment]
thank you so much i just spent hours trying to figure this out!!!

**Comment by hkropp** [comment]
This works! Thanks for sharing.

Important:
* append `.html`
* use full path `/a/b/[c]/d.html`

Simply copy what you find in `routes-manifest.json` :)

**Comment by jvsteiner** [comment]
Works for me - this thread/approach seems super under-rated.  I actually don't understand why this is not integrated into next already by default for static output targets.
